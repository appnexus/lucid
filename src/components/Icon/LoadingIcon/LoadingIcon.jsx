import React from 'react';
import Icon from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';
import { createClass } from '../../../util/component-types';

const { oneOf } = React.PropTypes;
const cx = lucidClassNames.bind('&-LoadingIcon');
const durations = {
	fast: '0.75s',
	normal: '1.25s',
	slow: '4s',
};

/**
 *
 * {"categories": ["visual design", "icons"], "extend": "Icon", "madeFrom": ["Icon"]}
 *
 * A loading icon.
 */
const LoadingIcon = createClass({
	displayName: 'LoadingIcon',
	propTypes: {
		...Icon.propTypes,
		/**
		 * The speed of rotation of the spinner. Defaults to 'normal'
		 */
		speed: oneOf(['fast', 'normal', 'slow']),
	},
	getDefaultProps() {
		return {
			...Icon.getDefaultProps(),
			speed: 'normal',
		};
	},
	render() {
		const {
			className,
			speed,
			style,
			...passThroughs,
		} = this.props;

		const animationDuration = `${durations[speed] || durations.normal}`;

		return (
			<Icon
				{...passThroughs}
				viewBox='0 0 100 100'
				className={cx('&', className)}
				style={{ animationDuration, ...style }}
			>
				<rect
					x='0'
					y='0'
					width='100'
					height='100'
					fill='none'
				/>
				<circle
					className={cx('&-circle')}
					cx='50'
					cy='50'
					r='40'
				/>
				<circle
					style={{ animationDuration }}
					className={cx('&-spinner')}
					cx='50'
					cy='50'
					r='40'
				/>

			</Icon>
		);
	},
});

export default LoadingIcon;
