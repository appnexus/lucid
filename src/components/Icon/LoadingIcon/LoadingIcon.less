.lucid-LoadingIcon {
	// simple rotation in IE11 because IE11 doesn't support svg animations
	-ms-animation: rotate-animation linear infinite;

	circle {
		fill: none;
		stroke-width: 12;
		stroke-linecap: round;
	}

	&-spinner {
		transform-origin: center;
		animation: dash-animation linear infinite;
		stroke: @color-primary;
		stroke-dasharray: 150px 100px;
	}

	&-circle {
		stroke: @color-gray;
	}

}

@keyframes rotate-animation {
	to {
		transform: rotateZ(360deg);
	}
}

@keyframes dash-animation {
	0% {
		stroke-dashoffset: 0px;
		stroke-dasharray: 151.327px 100px;
	}
	50% {
		stroke-dashoffset: -251.327px;
		stroke-dasharray: 10px 241.327px;
	}
	100% {
		stroke-dashoffset: -502.654px;
		stroke-dasharray: 151.327px 100px;
	}
}

