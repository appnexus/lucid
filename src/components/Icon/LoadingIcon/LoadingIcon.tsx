import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import Icon, { IIconProps } from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';

const cx = lucidClassNames.bind('&-LoadingIcon');

export enum durations {
	fast = '0.75s',
	normal = '1.25s',
	slow = '4s',
}

export interface ILoadingIconProps extends IIconProps {
	/**	The speed of rotation of the spinner. */
	speed?: keyof typeof durations;
}

export const loadingIconPropTypes = {
	/**	The speed of rotation of the spinner. */
	speed: PropTypes.oneOf(['fast', 'normal', 'slow']),

	/** Size variations of the icons. \`size\` directly effects height and width
		but the developer should also be conscious of the relationship with
		\`viewBox\`. */
	size: PropTypes.number,

	/** Size handles width and height, whereas \`width\` can manually override the width that would be set by size. */
	width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

	/** Size handles width and height, whereas \`height\` can manually override the height that would be set by size. */
	height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

	/** Any valid SVG aspect ratio. */
	aspectRatio: PropTypes.string,

	/** Adds styling that makes the icon appear clickable. */
	isClickable: PropTypes.bool,

	/** Adds styling that makes the icon appear disabled.  Also forces
		isClickable to be false. */
	isDisabled: PropTypes.bool,

	/** Called when the user clicks the \`Icon\`. Signature:
		\`({event, props}) => {}\` */
	onClick: PropTypes.func,

	/** Called when the user clicks an active, clickable \`Icon\`. Signature:
		\`({event, props}) => {}\` */
	onSelect: PropTypes.func,

	/** Any valid React children. */
	children: PropTypes.element,

	/** Classes that are appended to the component defaults. This prop is run
		through the \`classnames\` library. */
	className: PropTypes.string,
};

export const LoadingIcon = ({
	className,
	speed = 'normal',
	style,
	isDisabled,
	...passThroughs
}: ILoadingIconProps) => {
	const animationDuration = `${durations[speed] || durations.normal}`;

	return (
		<Icon
			{..._.omit(passThroughs, ['initialState', 'viewbox', 'color'])}
			viewBox='0 0 100 100'
			className={cx('&', className)}
			style={{ animationDuration, ...style }}
			isDisabled={isDisabled}
		>
			<rect x='0' y='0' width='100' height='100' fill='none' />
			<circle className={cx('&-circle')} cx='50' cy='50' r='40' />
			<circle
				style={{ animationDuration }}
				className={cx('&-spinner', { '&-spinner-is-disabled': isDisabled })}
				cx='50'
				cy='50'
				r='40'
			/>
		</Icon>
	);
};

LoadingIcon.displayName = 'LoadingIcon';

LoadingIcon.propTypes = loadingIconPropTypes;

LoadingIcon.defaultProps = Icon.defaultProps;

export default LoadingIcon;
