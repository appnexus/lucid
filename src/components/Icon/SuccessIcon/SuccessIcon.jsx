import _ from 'lodash';
import React from 'react';
import Icon from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';
import {
	createClass as createReactClass,
	omitProps,
} from '../../../util/component-types';

const cx = lucidClassNames.bind('&-SuccessIcon');

/**
 *
 * {"categories": ["visual design", "icons"], "extend": "CheckIcon", "madeFrom": ["CheckIcon"]}
 *
 * A success icon.
 */
const SuccessIcon = createReactClass({
	displayName: 'SuccessIcon',
	propTypes: {
		...Icon.propTypes,
	},

	render() {
		const { className, isDisabled, isClickable, ...passThroughs } = this.props;

		return (
			<Icon
				{...omitProps(passThroughs, SuccessIcon, [], false)}
				{..._.pick(passThroughs, _.keys(Icon.propTypes))}
				isClickable={isClickable}
				isDisabled={isDisabled}
				className={cx('&', className, isClickable && '&-is-clickable')}
			>
				<circle
					className={cx('&-background', {
						'&-background-is-disabled': isDisabled,
					})}
					cx="8"
					cy="8"
					r="8"
				/>
				<path
					className={cx('&-check')}
					d="M11.92,5.189L11.701,4.96c-0.285-0.285-0.75-0.285-1.033,0l-4.13,4.127L4.791,7.474 c-0.286-0.286-0.748-0.286-1.033,0L3.471,7.816c-0.286,0.286-0.286,0.746,0,1.033l2.485,2.354c0.285,0.285,0.748,0.285,1.033,0 l4.932-4.98C12.205,5.938,12.205,5.475,11.92,5.189z"
				/>
			</Icon>
		);
	},
});

export default SuccessIcon;
