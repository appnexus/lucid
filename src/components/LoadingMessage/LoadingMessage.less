.lucid-LoadingMessage {

	.lucid-Icon {
		display: inline-block;
	}

	@message-padding: @size-standard;

	line-height: 1em;
	background-color: @color-white;
	position: relative;
	display: inline-block;
	text-align: center;
	border: @border-lightBorder;
	border-radius: @size-borderRadius;
	padding: @message-padding;
	width: 135px + (@message-padding * 2);

	&-title {
		font: bold @size-L @fontFamily;
		margin: 6px 0 0;
	}

	&-body {
		font: @font;
		margin: 6px 0 0;
		display: inline-block;
	}
	&-close {
		position: absolute;
		top: 6px;
		right: 6px;
		cursor: pointer;
		fill: @color-gray-50;
		&:hover {
			fill: @color-black;
		}
	}
	&-no-content {
		width: inherit;
	}
}
