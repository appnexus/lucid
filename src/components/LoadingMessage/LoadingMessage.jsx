import _ from 'lodash';
import React from 'react';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass, findTypes } from '../../util/component-types';
import LoadingIcon from '../Icon/LoadingIcon/LoadingIcon';

const cx = lucidClassNames.bind('&-LoadingMessage');

const {
	node,
	string,
} = React.PropTypes;

/**
 *
 * {"categories": ["communication"], "madeFrom": ["LoadingIcon"]}
 *
 * A loading message.
 *
 */
const LoadingMessage = createClass({
	propName: 'LoadingMessage',
	displayName: 'LoadingMessage',
	_lucidIsPrivate: true,
	propTypes: {
		/**
		 * Class names that are appended to the defaults.
		 */
		className: string,
		/**
		 * Any valid React children.
		 */
		children: node,
	},

	components: {
		Icon: createClass({
			displayName: 'LoadingMessage.Icon',
			propName: 'Icon',
		}),
		Title: createClass({
			displayName: 'LoadingMessage.Title',
			propName: 'Title',
		}),
		Body: createClass({
			displayName: 'LoadingMessage.Body',
			propName: 'Body',
		}),
	},

	render() {
		const {
			props,
			props: {
				className,
				...passThroughs,
			},
		} = this;

		const {
			Icon,
			Title,
			Body,
		} = LoadingMessage;

		const defaultTitle = 'Loading';
		const iconElement = _.first(findTypes(props, Icon));
		const iconChildren = _.get(iconElement, 'props.children', <LoadingIcon />);
		const titleElement = _.first(findTypes(props, Title));
		const titleChildren = _.get(titleElement, 'props.children');
		const bodyElement = _.first(findTypes(props, Body));
		const bodyChildren = _.get(bodyElement, 'props.children', null);

		return (
			<div
				{..._.omit(passThroughs, ['Icon', 'Title', 'Body', 'children'])}
				className={cx('&', { '&-no-content': _.isNull(titleChildren) && !bodyChildren }, className)}
			>
				{iconChildren}
				{!_.isNull(titleChildren) && <h3 className={cx('&-title')}>{titleChildren || defaultTitle}</h3>}
				{bodyChildren && <span className={cx('&-body')}>{bodyChildren}</span>}
			</div>
		);
	},
});

export default LoadingMessage;
