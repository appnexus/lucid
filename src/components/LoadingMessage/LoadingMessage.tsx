/* eslint-disable react/prop-types */
import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import { getFirst, omitProps, StandardProps } from '../../util/component-types';
import LoadingIcon from '../Icon/LoadingIcon/LoadingIcon';

const cx = lucidClassNames.bind('&-LoadingMessage');

const { any, node, string } = PropTypes;

/** Loading Message Icon */
export interface ILoadingMessageIconProps extends StandardProps {
	description?: string;
}
const LoadingMessageIcon = (_props: ILoadingMessageIconProps): null => null;
LoadingMessageIcon.displayName = 'LoadingMessage.Icon';
LoadingMessageIcon.peek = {
	description: `Renders the \`Icon\` element passed in.`,
};
LoadingMessageIcon.propName = 'Icon';
LoadingMessageIcon.propTypes = {
	description: string,
	children: any,
};

/** Loading Message Title */
export interface ILoadingMessageTitleProps extends StandardProps {
	description?: string;
}
const LoadingMessageTitle = (_props: ILoadingMessageTitleProps): null => null;
LoadingMessageTitle.displayName = 'LoadingMessage.Title';
LoadingMessageTitle.peek = {
	description: `Renders an \`<h3>\` that represents the title of the \`LoadingMessage\`.  Defaults to the string "Loading".`,
};
LoadingMessageTitle.propName = 'Title';
LoadingMessageTitle.propTypes = {
	description: string,
	children: any,
};

/** Loading Message Body */
export interface ILoadingMessageBodyProps extends StandardProps {
	description?: string;
}
const LoadingMessageBody = (_props: ILoadingMessageBodyProps): null => null;
LoadingMessageBody.displayName = 'LoadingMessage.Body';
LoadingMessageBody.peek = {
	description: `Renders a \`<span>\` that represents the body of the \`LoadingMessage\`.`,
};
LoadingMessageBody.propName = 'Body';
LoadingMessageBody.propTypes = {
	description: string,
	children: any,
};

/** Loading Message */
export interface ILoadingMessageProps
	extends StandardProps,
		React.DetailedHTMLProps<
			React.HTMLAttributes<HTMLDivElement>,
			HTMLDivElement
		> {
	/** Custom Icon element (alias for `LoadingMessage.Icon`) */
	Icon?: React.ReactNode;

	/** Custom Title element (alias for `LoadingMessage.Title`) */
	Title?: React.ReactNode;

	/** Custom Body element (alias for `LoadingMessage.Body`) */
	Body?: React.ReactNode;
}

export const LoadingMessage = (
	props: ILoadingMessageProps
): React.ReactElement => {
	const { className, ...passThroughs } = props;
	const { Icon, Title, Body } = LoadingMessage;

	const defaultTitle = 'Loading';
	const iconElement = getFirst(props, Icon);
	const iconChildren = _.get(iconElement, 'props.children', <LoadingIcon />);
	const titleElement = getFirst(props, Title);
	const titleChildren = _.get(titleElement, 'props.children');
	const bodyElement = getFirst(props, Body);
	const bodyChildren = _.get(bodyElement, 'props.children', null);

	return (
		<div
			{...omitProps(passThroughs, undefined, _.keys(LoadingMessage.propTypes))}
			className={cx(
				'&',
				{ '&-no-content': _.isNull(titleChildren) && !bodyChildren },
				className
			)}
		>
			{iconChildren}
			{!_.isNull(titleChildren) && (
				<h3 className={cx('&-title')}>{titleChildren || defaultTitle}</h3>
			)}
			{bodyChildren && <span className={cx('&-body')}>{bodyChildren}</span>}
		</div>
	);
};

LoadingMessage.Icon = LoadingMessageIcon;
LoadingMessage.Title = LoadingMessageTitle;
LoadingMessage.Body = LoadingMessageBody;

LoadingMessage._isPrivate = true;
LoadingMessage.displayName = 'LoadingMessage';
LoadingMessage.peek = {
	description: `A loading message.`,
	categories: ['communication'],
	madeFrom: ['LoadingIcon'],
};
LoadingMessage.propTypes = {
	/**
		Class names that are appended to the defaults.
	*/
	className: string,

	/**
		Any valid React children.
	*/
	children: node,

	/**
		Custom Icon element (alias for \`LoadingMessage.Icon\`)
	*/
	Icon: node,

	/**
		Custom Title element (alias for \`LoadingMessage.Title\`)
	*/
	Title: node,

	/**
		Custom Body element (alias for \`LoadingMessage.Body\`)
	*/
	Body: node,
};

export default LoadingMessage;
