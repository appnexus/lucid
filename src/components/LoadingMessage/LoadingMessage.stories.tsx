import _ from 'lodash';
import React from 'react';
import { Story, Meta } from '@storybook/react';

import LoadingMessage, { ILoadingMessageProps } from './LoadingMessage';

export default {
	title: 'Loading Message/LoadingMessage',
	component: LoadingMessage,
	parameters: {
		docs: {
			description: {
				component: LoadingMessage.peek.description,
			},
		},
		layout: 'centered',
	},
} as Meta;

export const Basic: Story<ILoadingMessageProps> = (args) => {
	return <LoadingMessage />;
};

export const CustomTitle: Story<ILoadingMessageProps> = (args) => {
	return <LoadingMessage {...args} />;
};
CustomTitle.args = {
	Title: 'A custom title',
};

export const CustomBody: Story<ILoadingMessageProps> = (args) => {
	return <LoadingMessage {...args} />;
};
CustomBody.args = {
	Body: 'A custom body',
};
