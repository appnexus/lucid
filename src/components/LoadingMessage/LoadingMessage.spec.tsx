import React from 'react';
import { shallow } from 'enzyme';
import assert from 'assert';
import { common } from '../../util/generic-tests';
import LoadingMessage from './LoadingMessage';
import LoadingIcon from '../Icon/LoadingIcon/LoadingIcon';
import SuccessIcon from '../Icon/SuccessIcon/SuccessIcon';

const { Icon, Title, Body } = LoadingMessage;

describe('LoadingMessage', () => {
	common(LoadingMessage);

	describe('render', () => {
		let wrapper;

		beforeEach(() => (wrapper = shallow(<LoadingMessage />)));

		it('should render a LoadingIcon', () => {
			assert.equal(
				wrapper.find(LoadingIcon).length,
				1,
				'must render loading icon'
			);
		});

		it('should render "Loading" as the title', () => {
			assert.equal(
				wrapper.find('.lucid-LoadingMessage-title').text(),
				'Loading'
			);
		});

		it('should not render a body', () => {
			assert.equal(wrapper.find('.lucid-LoadingMessage-body').length, 0);
		});

		it('should not have `&-no-content` class', () => {
			assert(!wrapper.hasClass('lucid-LoadingMessage-no-content'));
		});
	});

	describe('child components', () => {
		describe('Icon', () => {
			it('should render the custom icon', () => {
				const wrapper = shallow(
					<LoadingMessage>
						<Icon>
							<SuccessIcon />
						</Icon>
					</LoadingMessage>
				);
				assert.equal(wrapper.find(SuccessIcon).length, 1);
			});

			it('should render the custom icon', () => {
				const wrapper = shallow(<LoadingMessage Icon={<SuccessIcon />} />);
				assert.equal(wrapper.find(SuccessIcon).length, 1);
			});
		});

		describe('Title', () => {
			it('should render the custom title', () => {
				const wrapper = shallow(
					<LoadingMessage>
						<Title>A custom title</Title>
					</LoadingMessage>
				);
				assert.equal(
					wrapper.find('.lucid-LoadingMessage-title').text(),
					'A custom title'
				);
			});

			it('should render the custom title', () => {
				const wrapper = shallow(<LoadingMessage Title='A custom title' />);
				assert.equal(
					wrapper.find('.lucid-LoadingMessage-title').text(),
					'A custom title'
				);
			});

			describe('null', () => {
				let wrapper;

				beforeEach(() => (wrapper = shallow(<LoadingMessage Title={null} />)));

				it('should render a LoadingIcon', () => {
					assert.equal(
						wrapper.find(LoadingIcon).length,
						1,
						'must render loading icon'
					);
				});

				it('should not render a title', () => {
					assert.equal(wrapper.find('.lucid-LoadingMessage-title').length, 0);
				});

				it('should add `&-no-content` class', () => {
					assert(wrapper.hasClass('lucid-LoadingMessage-no-content'));
				});
			});
		});

		describe('Body', () => {
			it('should render the custom body', () => {
				const wrapper = shallow(
					<LoadingMessage>
						<Body>A custom body</Body>
					</LoadingMessage>
				);
				assert.equal(
					wrapper.find('.lucid-LoadingMessage-body').text(),
					'A custom body'
				);
			});

			it('should render the custom body', () => {
				const wrapper = shallow(<LoadingMessage Body='A custom body' />);
				assert.equal(
					wrapper.find('.lucid-LoadingMessage-body').text(),
					'A custom body'
				);
			});
		});
	});
});
