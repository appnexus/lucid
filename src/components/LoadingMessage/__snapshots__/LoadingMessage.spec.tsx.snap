// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingMessage [common] example testing should match snapshot(s) for Basic 1`] = `
<div
  class="lucid-LoadingMessage"
>
  <svg
    class="lucid-Icon lucid-Icon-color-primary lucid-LoadingIcon"
    height="16"
    preserveAspectRatio="xMidYMid meet"
    style="animation-duration:1.25s"
    viewBox="0 0 100 100"
    width="16"
  >
    <rect
      fill="none"
      height="100"
      width="100"
      x="0"
      y="0"
    />
    <circle
      class="lucid-LoadingIcon-circle"
      cx="50"
      cy="50"
      r="40"
    />
    <circle
      class="lucid-LoadingIcon-spinner"
      cx="50"
      cy="50"
      r="40"
      style="animation-duration:1.25s"
    />
  </svg>
  <h3
    class="lucid-LoadingMessage-title"
  >
    Loading
  </h3>
</div>
`;

exports[`LoadingMessage [common] example testing should match snapshot(s) for CustomBody 1`] = `
<div
  class="lucid-LoadingMessage"
>
  <svg
    class="lucid-Icon lucid-Icon-color-primary lucid-LoadingIcon"
    height="16"
    preserveAspectRatio="xMidYMid meet"
    style="animation-duration:1.25s"
    viewBox="0 0 100 100"
    width="16"
  >
    <rect
      fill="none"
      height="100"
      width="100"
      x="0"
      y="0"
    />
    <circle
      class="lucid-LoadingIcon-circle"
      cx="50"
      cy="50"
      r="40"
    />
    <circle
      class="lucid-LoadingIcon-spinner"
      cx="50"
      cy="50"
      r="40"
      style="animation-duration:1.25s"
    />
  </svg>
  <h3
    class="lucid-LoadingMessage-title"
  >
    Loading
  </h3>
  <span
    class="lucid-LoadingMessage-body"
  >
    A custom body
  </span>
</div>
`;

exports[`LoadingMessage [common] example testing should match snapshot(s) for CustomTitle 1`] = `
<div
  class="lucid-LoadingMessage"
>
  <svg
    class="lucid-Icon lucid-Icon-color-primary lucid-LoadingIcon"
    height="16"
    preserveAspectRatio="xMidYMid meet"
    style="animation-duration:1.25s"
    viewBox="0 0 100 100"
    width="16"
  >
    <rect
      fill="none"
      height="100"
      width="100"
      x="0"
      y="0"
    />
    <circle
      class="lucid-LoadingIcon-circle"
      cx="50"
      cy="50"
      r="40"
    />
    <circle
      class="lucid-LoadingIcon-spinner"
      cx="50"
      cy="50"
      r="40"
      style="animation-duration:1.25s"
    />
  </svg>
  <h3
    class="lucid-LoadingMessage-title"
  >
    A custom title
  </h3>
</div>
`;
